cmd/kubeadm/app/cmd/token_test.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"flag"
+	"regexp"
+	"testing"
+
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+const (
+	TokenExpectedRegex = "^\\S{6}\\.\\S{16}\n$"
+)
+
+var kubeadmPath string
+
+func init() {
+	flag.StringVar(&kubeadmPath, "kubeadm-path", "", "Location of kubeadm")
```

Comment:
TL;DR: I'd like to continue categorizing these tests as `cmd` tests, have them run during the `make test-cmd` target, and create a new tree for them to live in, `kubernetes/test/cmd`.

So, there's a lot of discussion around exactly what kind of tests these are and where they should live. I think there are a few dimensions at play:

- What to call them
- When they should execute
- Where the code should live

To my understanding, we currently have four kinds of tests, with different Makefile targets for executing each kind:

| Target | Description |
| --- | --- |
| make test | Unit tests |
| make test-integration | Integration tests |
| make test-cmd | CLI tests |
| make test-e2e | End-to-end tests |

(From [build-tools/README.md](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/build-tools/README.md) and [Makefile](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/Makefile#L116))

In spirit, these tests fit in very nicely with the existing `test-cmd` CLI tests that exist to test `kubectl`, `kubelet`, `kube-apiserver`, etc. Like those tests, mine require that the binaries be built before the tests can run, and they actually test the functionality at the CLI level to catch things that unit tests might miss (or can't test), just like the existing `test-cmd` tests.

The only reason I didn't add them to the existing `test-cmd` tests is because 1) all of those tests are in an enormous monolithic file that makes it difficult to selectively run subsets of tests, 2) the file is all bash, and 3) it seemed counter-productive to pulling `kubeadm` out to its own repo in the future. (Here is that file for reference: [test-cmd.sh](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/hack/make-rules/test-cmd.sh))

If it weren't for those things, I would have just added these tests directly to `test-cmd.sh` and I don't think there would have been much to discuss. So to answer the question of "what kinds of tests they are," I would call them `cmd` tests. I don't think anyone is arguing that they're unit tests, and they're far from what we call end-to-end tests (which actually spin up an entire cluster for testing). It's debatable if they should be called integration tests, but our existing integration tests don't test at the CLI level, and moreover, these tests aren't really testing the integration of multiple components, they're just testing the same simple component at a higher level. We may absolutely have kubeadm integration tests in the future, but I don't think these really qualify in spirit or match the expectations of our existing `make test-integration` target. So, I would like to continue calling them `cmd` tests and having them execute as part of `make test-cmd`.

As for where the tests should actually live in the codebase, I originally had them in a separate test file in the `cmd/kubeadm/app` tree, alongside the unit test file, and only merged them into the unit test file upon request. I do, however, think that there's merit to separating out the different kinds of tests we have at the file level, as all of our existing test types do (unit test files are stored right alongside the implementing code, integration tests have their own tree, and `cmd` and `e2e` tests are completely separate in the `hack` tree). That makes it easy to find all of the unit tests in the same place, and understand how to add a unit test by following the same pattern, and the same goes for integration tests, etc. If we merge multiple different kinds of tests into the same file, it makes it easier to shoot yourself in the foot and either have your test run as part of the wrong make test target, have it run redundantly in multiple make test targets, or even worse, accidentally be run in none of our make test targets and silently get ignored because you followed the wrong convention and didn't realize it never ran. I think having different kinds of tests in the same file is just asking for trouble.

So, what I'd propose is to create a new tree in `test/` akin to the existing `test/integration` and `test/e2e` trees: `test/cmd`. This keeps it separate from the other test types and makes it very easy to find all of the `cmd` tests to run during the `make test-cmd` target. I assume the only reason this doesn't exist right now is because of the fact that all of the existing `cmd` tests were written in a single bash file, so there was no need for further organization or hierarchy.